File: apps/ios/Janark/UI/Shell/JanarkScaffold.swift
import SwiftUI

struct NavDest: Identifiable {
    let route: String
    let label: String
    let systemImage: String
    let group: String
    var id: String { route }
}

private let drawerItems: [NavDest] = [
    NavDest(route: Routes.home, label: "Home", systemImage: "house", group: "Discover"),
    NavDest(route: Routes.activity, label: "Pulse", systemImage: "chart.bar", group: "Discover"),
    NavDest(route: Routes.issues, label: "Issues", systemImage: "flame", group: "Act"),
    NavDest(route: Routes.petitions, label: "Petitions", systemImage: "megaphone", group: "Act"),
    NavDest(route: Routes.reports, label: "Reports", systemImage: "exclamationmark.bubble", group: "Act"),
    NavDest(route: Routes.votes, label: "Votes", systemImage: "checkmark.seal", group: "Act"),
    NavDest(route: Routes.discussions, label: "Discussions", systemImage: "bubble.left.and.bubble.right", group: "Act"),
    NavDest(route: Routes.notices, label: "Notices", systemImage: "megaphone", group: "Act"),
    NavDest(route: Routes.profile, label: "Profile", systemImage: "person.circle", group: "You"),
    NavDest(route: Routes.settings, label: "Settings", systemImage: "gearshape", group: "You"),
    NavDest(route: Routes.about, label: "About", systemImage: "info.circle", group: "You"),
]

private let tabs: [NavDest] = [
    NavDest(route: Routes.home, label: "Home", systemImage: "house", group: "Discover"),
    NavDest(route: Routes.activity, label: "Pulse", systemImage: "chart.bar", group: "Discover"),
    NavDest(route: Routes.petitions, label: "Petitions", systemImage: "megaphone", group: "Act"),
    NavDest(route: Routes.profile, label: "Profile", systemImage: "person.circle", group: "You"),
]

struct JanarkScaffold<Content: View>: View {
    @EnvironmentObject private var app: AppContainer
    @Environment(\.janarkColors) private var c

    let currentRoute: String?
    let isCompose: Bool
    let onNavigate: (String) -> Void
    @ViewBuilder let content: Content

    @State private var showMenu = false
    @State private var showCreate = false

    private var selectedTab: String? {
        tabs.first(where: { currentRoute == $0.route || (currentRoute?.hasPrefix($0.route) ?? false) })?.route
    }

    var body: some View {
        ZStack {
            VStack(spacing: 0) {
                topBar
                content
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
                if !isCompose {
                    bottomBar
                }
            }
            .background(c.background)

            if !isCompose {
                VStack {
                    Spacer()
                    HStack {
                        Spacer()
                        Button {
                            showCreate = true
                        } label: {
                            Image(systemName: "plus")
                                .font(.title2.weight(.semibold))
                                .foregroundStyle(c.chrome)
                                .frame(width: 56, height: 56)
                                .background(c.amber, in: Circle())
                                .shadow(color: .black.opacity(0.18), radius: 8, y: 3)
                        }
                        .padding(.trailing, 20)
                        .padding(.bottom, 72)
                    }
                }
            }

            if showMenu {
                drawerOverlay
            }
        }
        .sheet(isPresented: $showCreate) {
            createSheet
                .presentationDetents([.medium, .large])
        }
    }

    private var topBar: some View {
        HStack(spacing: 10) {
            Button { showMenu = true } label: {
                Image(systemName: "line.3.horizontal")
                    .foregroundStyle(c.onChrome)
                    .frame(width: 36, height: 36)
            }
            Image("JanarkLogo")
                .resizable()
                .scaledToFit()
                .frame(width: 28, height: 28)
            Text("Janark")
                .font(.system(.title3, design: .serif).weight(.semibold))
                .foregroundStyle(c.onChrome)
            Spacer()
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 10)
        .background(c.chrome)
    }

    private var bottomBar: some View {
        HStack {
            ForEach(tabs) { tab in
                Button {
                    onNavigate(tab.route)
                } label: {
                    VStack(spacing: 4) {
                        Image(systemName: tab.systemImage)
                        Text(tab.label)
                            .font(.caption2)
                    }
                    .foregroundStyle(selectedTab == tab.route ? c.amber : c.onChrome.opacity(0.7))
                    .frame(maxWidth: .infinity)
                }
                .buttonStyle(.plain)
            }
        }
        .padding(.top, 10)
        .padding(.bottom, 8)
        .background(c.chrome)
    }

    private var drawerOverlay: some View {
        ZStack(alignment: .leading) {
            Color.black.opacity(0.35)
                .ignoresSafeArea()
                .onTapGesture { showMenu = false }

            VStack(alignment: .leading, spacing: 0) {
                HStack(spacing: 10) {
                    Image("JanarkLogo")
                        .resizable()
                        .scaledToFit()
                        .frame(width: 40, height: 40)
                    VStack(alignment: .leading, spacing: 2) {
                        Text("Janark")
                            .font(.system(.title2, design: .serif).weight(.semibold))
                            .foregroundStyle(c.foreground)
                        Text(app.session.uiSession.anonId ?? "Browse freely")
                            .font(.caption)
                            .foregroundStyle(c.muted)
                    }
                }
                .padding(16)

                ScrollView {
                    let groups = Dictionary(grouping: drawerItems, by: \.group)
                    ForEach(["Discover", "Act", "You"], id: \.self) { group in
                        if let items = groups[group] {
                            SectionLabel(group)
                                .padding(.horizontal, 16)
                            ForEach(items) { item in
                                Button {
                                    onNavigate(item.route)
                                    showMenu = false
                                } label: {
                                    HStack(spacing: 12) {
                                        Image(systemName: item.systemImage)
                                        Text(item.label)
                                        Spacer()
                                    }
                                    .foregroundStyle(currentRoute == item.route ? c.amber : c.foreground)
                                    .padding(.horizontal, 16)
                                    .padding(.vertical, 12)
                                    .background(currentRoute == item.route ? c.amber.opacity(0.12) : Color.clear)
                                }
                                .buttonStyle(.plain)
                            }
                        }
                    }
                }
                Spacer(minLength: 0)
            }
            .frame(width: 300)
            .frame(maxHeight: .infinity)
            .background(c.surface)
            .transition(.move(edge: .leading))
        }
    }

    private var createSheet: some View {
        NavigationStack {
            List {
                Section {
                    createRow(CreateKind.share, title: "Share with community", reason: "share with your community")
                    createRow(CreateKind.report, title: "File a report", reason: "file a report")
                } header: {
                    Text("Quick")
                }
                Section {
                    ForEach([CreateKind.issue, .petition, .vote, .discussion, .notice, .meme]) { kind in
                        createRow(kind, title: kind.title, reason: kind.title.lowercased())
                    }
                } header: {
                    Text("Organize")
                }
            }
            .navigationTitle("Create")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) {
                    Button("Close") { showCreate = false }
                }
            }
        }
    }

    private func createRow(_ kind: CreateKind, title: String, reason: String) -> some View {
        Button(title) {
            showCreate = false
            let dest = Routes.create(kind.path)
            if app.session.uiSession.authenticated {
                onNavigate(dest)
            } else {
                app.auth.require(reason) { onNavigate(dest) }
            }
        }
    }
}

File: apps/android/app/src/main/java/org/janark/app/ui/shell/JanarkScaffold.kt
package org.janark.app.ui.shell

import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.outlined.AccountCircle
import androidx.compose.material.icons.outlined.Campaign
import androidx.compose.material.icons.outlined.ChatBubbleOutline
import androidx.compose.material.icons.outlined.Home
import androidx.compose.material.icons.outlined.HowToVote
import androidx.compose.material.icons.outlined.Image
import androidx.compose.material.icons.outlined.Info
import androidx.compose.material.icons.outlined.LocalFireDepartment
import androidx.compose.material.icons.outlined.Menu
import androidx.compose.material.icons.outlined.Place
import androidx.compose.material.icons.outlined.Report
import androidx.compose.material.icons.outlined.Settings
import androidx.compose.material3.DrawerValue
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.FloatingActionButton
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.ModalNavigationDrawer
import androidx.compose.material3.NavigationBar
import androidx.compose.material3.NavigationBarItem
import androidx.compose.material3.NavigationBarItemDefaults
import androidx.compose.material3.NavigationDrawerItem
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.material3.rememberDrawerState
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.launch
import org.janark.app.R
import org.janark.app.ui.LocalApp
import org.janark.app.ui.components.SectionLabel
import org.janark.app.ui.nav.CreateKind
import org.janark.app.ui.nav.Routes
import org.janark.app.ui.theme.JanarkTheme
import org.janark.app.ui.theme.toC

data class NavDest(val route: String, val label: String, val icon: ImageVector, val group: String)

private val drawerItems = listOf(
    NavDest(Routes.Home, "Home", Icons.Outlined.Home, "Discover"),
    NavDest(Routes.Activity, "Pulse", Icons.Outlined.Place, "Discover"),
    NavDest(Routes.Issues, "Issues", Icons.Outlined.LocalFireDepartment, "Act"),
    NavDest(Routes.Petitions, "Petitions", Icons.Outlined.Campaign, "Act"),
    NavDest(Routes.Reports, "Reports", Icons.Outlined.Report, "Act"),
    NavDest(Routes.Votes, "Votes", Icons.Outlined.HowToVote, "Act"),
    NavDest(Routes.Discussions, "Discussions", Icons.Outlined.ChatBubbleOutline, "Act"),
    NavDest(Routes.Notices, "Notices", Icons.Outlined.Campaign, "Act"),
    NavDest(Routes.Profile, "Profile", Icons.Outlined.AccountCircle, "You"),
    NavDest(Routes.Settings, "Settings", Icons.Outlined.Settings, "You"),
    NavDest(Routes.About, "About", Icons.Outlined.Info, "You"),
)

private val tabs = listOf(
    NavDest(Routes.Home, "Home", Icons.Outlined.Home, "Discover"),
    NavDest(Routes.Activity, "Pulse", Icons.Outlined.Place, "Discover"),
    NavDest(Routes.Petitions, "Petitions", Icons.Outlined.Campaign, "Act"),
    NavDest(Routes.Profile, "Profile", Icons.Outlined.AccountCircle, "You"),
)

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun JanarkScaffold(
    currentRoute: String?,
    isCompose: Boolean,
    onNavigate: (String) -> Unit,
    content: @Composable () -> Unit,
) {
    val c = JanarkTheme.colors
    val drawer = rememberDrawerState(DrawerValue.Closed)
    val scope = rememberCoroutineScope()
    var createOpen by remember { mutableStateOf(false) }
    val app = LocalApp.current
    val session by app.session.uiSession.collectAsState(initial = org.janark.app.data.session.UiSession())

    val selectedTab = tabs.firstOrNull { currentRoute == it.route || currentRoute?.startsWith(it.route) == true }?.route

    ModalNavigationDrawer(
        drawerState = drawer,
        drawerContent = {
            Column(
                Modifier
                    .fillMaxHeight()
                    .width(300.dp)
                    .padding(16.dp),
            ) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Image(painterResource(R.drawable.janark_logo), null, Modifier.size(40.dp))
                    Spacer(Modifier.width(10.dp))
                    Column {
                        Text("Janark", fontFamily = FontFamily.Serif, fontWeight = FontWeight.SemiBold, fontSize = 22.sp, color = c.foreground.toC())
                        Text(session.anonId ?: "Browse freely", color = c.muted.toC(), fontSize = 12.sp)
                    }
                }
                Spacer(Modifier.height(16.dp))
                drawerItems.groupBy { it.group }.forEach { (group, items) ->
                    SectionLabel(group)
                    items.forEach { item ->
                        NavigationDrawerItem(
                            label = { Text(item.label) },
                            selected = currentRoute == item.route,
                            onClick = {
                                onNavigate(item.route)
                                scope.launch { drawer.close() }
                            },
                            icon = { Icon(item.icon, null) },
                        )
                    }
                }
            }
        },
    ) {
        Scaffold(
            containerColor = c.background.toC(),
            topBar = {
                TopAppBar(
                    title = {
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Image(painterResource(R.drawable.janark_logo), null, Modifier.size(28.dp))
                            Spacer(Modifier.width(8.dp))
                            Text("Janark", fontFamily = FontFamily.Serif, fontWeight = FontWeight.SemiBold, color = c.onChrome.toC())
                        }
                    },
                    navigationIcon = {
                        IconButton(onClick = { scope.launch { drawer.open() } }) {
                            Icon(Icons.Outlined.Menu, contentDescription = "Menu", tint = c.onChrome.toC())
                        }
                    },
                    colors = TopAppBarDefaults.topAppBarColors(
                        containerColor = c.chrome.toC(),
                        titleContentColor = c.onChrome.toC(),
                    ),
                )
            },
            bottomBar = {
                if (!isCompose) {
                    NavigationBar(containerColor = c.chrome.toC()) {
                        tabs.forEach { tab ->
                            NavigationBarItem(
                                selected = selectedTab == tab.route,
                                onClick = { onNavigate(tab.route) },
                                icon = { Icon(tab.icon, tab.label) },
                                label = { Text(tab.label) },
                                colors = NavigationBarItemDefaults.colors(
                                    selectedIconColor = c.amber.toC(),
                                    selectedTextColor = c.amber.toC(),
                                    unselectedIconColor = c.onChrome.toC().copy(alpha = 0.7f),
                                    unselectedTextColor = c.onChrome.toC().copy(alpha = 0.7f),
                                    indicatorColor = c.chromeMid.toC(),
                                ),
                            )
                        }
                    }
                }
            },
            floatingActionButton = {
                if (!isCompose) {
                    FloatingActionButton(
                        onClick = { createOpen = true },
                        containerColor = c.amber.toC(),
                        contentColor = c.chrome.toC(),
                    ) { Icon(Icons.Filled.Add, contentDescription = "Create") }
                }
            },
        ) { padding ->
            Column(Modifier.padding(padding).fillMaxSize()) { content() }
        }
    }

    if (createOpen) {
        ModalBottomSheet(onDismissRequest = { createOpen = false }, sheetState = rememberModalBottomSheetState()) {
            Column(Modifier.padding(20.dp)) {
                SectionLabel("Quick")
                CreateKind.Share.let { k ->
                    NavigationDrawerItem(label = { Text("Share with community") }, selected = false, onClick = {
                        createOpen = false
                        if (session.authenticated) onNavigate(Routes.create(k.path))
                        else app.auth.require("share with your community") { onNavigate(Routes.create(k.path)) }
                    })
                }
                CreateKind.Report.let { k ->
                    NavigationDrawerItem(label = { Text("File a report") }, selected = false, onClick = {
                        createOpen = false
                        if (session.authenticated) onNavigate(Routes.create(k.path))
                        else app.auth.require("file a report") { onNavigate(Routes.create(k.path)) }
                    })
                }
                HorizontalDivider(Modifier.padding(vertical = 8.dp))
                SectionLabel("Organize")
                listOf(CreateKind.Issue, CreateKind.Petition, CreateKind.Vote, CreateKind.Discussion, CreateKind.Notice, CreateKind.Meme).forEach { k ->
                    NavigationDrawerItem(label = { Text(k.title) }, selected = false, onClick = {
                        createOpen = false
                        if (session.authenticated) onNavigate(Routes.create(k.path))
                        else app.auth.require(k.title.lowercase()) { onNavigate(Routes.create(k.path)) }
                    })
                }
                Spacer(Modifier.height(12.dp))
            }
        }
    }
}
